Marlin/src/feature/easythreed_ui.cpp

```diff
@@ -0,0 +1,242 @@
+              }
+              card.ls();                                            // List all files to serial output
+              const uint16_t filecnt = card.countFilesInWorkDir();  // Count printable files in cwd
+              if (filecnt == 0) return;                             // None are printable?
+              card.selectFileByIndex(filecnt);                      // Select the last file according to current sort options
+              card.openAndPrintFile(card.filename);                 // Start printing it
+              blink_interval_ms = LED_BLINK_2;                      // Blink the indicator LED at 1 second intervals
+              print_flag = true;                                    // Tell EasythreedUI printing should be active
+              print_key_flag = PF_PAUSE;                            // The "Print" button now pauses the print
+            }
+            break;
+          case PF_PAUSE:                                            // Pause printing (not currently firing)
+            blink_interval_ms = LED_ON;                             // Set indicator to steady ON
+            card.pauseSDPrint();                                    // Unceremoniously pause the SD print in progress
+            print_key_flag = PF_RESUME;                             // The "Print" button now resumes the print
+            break;
+          case PF_RESUME:                                           // Resume printing (not currently firing)
+            blink_interval_ms = LED_BLINK_2;                        // Blink the indicator LED at 1 second intervals
+            card.startOrResumeFilePrinting();                       // Unceremoniously resume the SD print in progress
+            print_key_flag = PF_PAUSE;                              // The "Print" button now pauses the print
+            break;
+        }
+      }
+      else {                                                        // Register a longer press
+        if (print_key_flag == PF_START) {                           // While not printing, this moves Z up 10mm
+          blink_interval_ms = LED_ON;
+          queue.inject(F("G91\nG0 Z10 F600\nG90"));                 // Raise Z soon after returning to main loop
+        }
+        else {                                                      // While printing, cancel print
+          wait_for_heatup = false;                                  // Break out of any wait-for-heating loop
+          quickstop_stepper();                                      // Hard-stop all steppers
+          thermalManager.disable_all_heaters();                     // Turn off all heaters
+          print_flag = false;
+          blink_interval_ms = LED_OFF;                              // Turn off LED
+        }
+        planner.synchronize();                                      // Wait for commands already in the planner to finish
+        TERN_(HAS_STEPPER_RESET, disableStepperDrivers());          // Disable all steppers - now!
```

Comment:
Note that the commands sent by `inject` above will be executed later, after exiting this method.

Reply:
Equivalent to… "G91\nG0 Z10 F600\nG90\nM400\nM18"